src/autopas/tuning/AutoTuner.cpp

```diff
@@ -0,0 +1,413 @@
+    if (getCurrentNumSamples() == _maxSamples) {
+      const long reducedValue = estimateRuntimeFromSamples();
+      _evidenceCollection.addEvidence(currentConfig, {_iteration, _tuningPhase, reducedValue});
+
+      // smooth evidence to remove high outliers. If smoothing results in a higher value use the original value.
+      const auto smoothedValue =
+          std::min(reducedValue, smoothing::smoothLastPoint(*_evidenceCollection.getEvidence(currentConfig), 5));
+
+      // replace collected evidence with smoothed value to improve next smoothing
+      _evidenceCollection.modifyLastEvidence(currentConfig).value = smoothedValue;
+
+      std::for_each(_tuningStrategies.begin(), _tuningStrategies.end(), [&](auto &tuningStrategy) {
+        tuningStrategy->addEvidence(getCurrentConfig(), _evidenceCollection.modifyLastEvidence(currentConfig));
+      });
+
+      // print config, times and reduced value
+      AutoPasLog(
+          DEBUG, "Collected {} for {}",
+          [&]() {
+            switch (this->_tuningMetric) {
+              case TuningMetricOption::time:
+                return "times";
+              case TuningMetricOption::energy:
+                return "energy consumption";
+            }
+            return "unknown tuning metric";
+          }(),
+          [&]() {
+            std::ostringstream ss;
+            // print config
+            ss << currentConfig << " : ";
+            // print all timings
+            ss << utils::ArrayUtils::to_string(_samplesRebuildingNeighborLists, " ",
+                                               {"With rebuilding neighbor lists [ ", " ] "});
+            ss << utils::ArrayUtils::to_string(_samplesNotRebuildingNeighborLists, " ",
+                                               {"Without rebuilding neighbor lists [ ", " ] "});
+            ss << "Smoothed value: " << smoothedValue;
+            return ss.str();
+          }());
+
+      _tuningDataLogger.logTuningData(currentConfig, _samplesRebuildingNeighborLists,
+                                      _samplesNotRebuildingNeighborLists, _iteration, reducedValue, smoothedValue);
+    }
+  }
+}
+
+void AutoTuner::bumpIterationCounters() {
+  ++_iteration;
+  ++_iterationsSinceTuning;
+  // this will NOT catch the first tuning phase because _iterationsSinceTuning is initialized to _tuningInterval.
+  // Hence, _tuningPhase is initialized as 1.
+  if (_iterationsSinceTuning == _tuningInterval) {
+    ++_tuningPhase;
+  }
+}
+
+bool AutoTuner::willRebuildNeighborLists() const {
+  const bool inTuningPhase = _iterationsSinceTuning >= _tuningInterval;
+  // How many iterations ago did the rhythm of rebuilds change?
+  const auto iterationBaseline = inTuningPhase ? (_iterationsSinceTuning - _tuningInterval) : _iterationsSinceTuning;
+  // What is the rebuild rhythm?
+  const auto iterationsPerRebuild = inTuningPhase ? _maxSamples : _rebuildFrequency;
+  return (iterationBaseline % iterationsPerRebuild) == 0;
+}
```

Comment:
Yeah, this thing could be more neat. I think I chose "rhythm" because I wanted to emphasize that the frequency in question here is not necessarily `_rebuildFrequency`. Feel free to make a better-fitting suggestion.

Right, but this begs the question of when we would collect fewer samples than `_maxSamples`. The only case I can think of (currently) is when a retune is triggered through a resize but in that case the logic should work because it always seeks the start of the tuning phase which is also updated.

Reply:
> My concern is that when this no longer becomes the case, this error is overlooked.

Very true

On the other hand, my concern with such a flag is that the AutoTuner gains more state that needs to be kept consistent. Problems with your suggestion that immediately jump to my mind are:
1. Is it enough to set the flag in `getNextConfig()`? Do we maybe need this information already earlier?
2. If `willRebuildNeighborListsDueToChangingConfiguration` resets the flag, we can't obtain the information multiple times.
3. So, where should we reset the flag? The AutoTuner doesn't (and shouldn't, for compilation reasons) have access to the container. Hence, it would not be notified about any change to it.

Overall, I agree that something should change, but maybe we can defer this to a later point (-> Issue) so we can merge these changes.